### PR TITLE
#5592: Optimize Falcon 7b lm head matmul

### PR DESCRIPTION
**Pull request by s-jovic**
Add a function for Falcon 7b lm head with seq len > 512. The function splits matmul into 4 or 8 2d matmuls, and expects the weights to be sliced accordingly. 

For 2k seq len, kernel time went from 400ms to 17ms (total kernel time for all slices and concats). For 1k seq len kernel time went from 200ms to 9ms. Other 32-multiple sequence lengths between 512 and 2048 are also supported.

The function is currently used only in the test - the idea is to incorporate it to the model with other pending optimizations; then some more optimizations and refactoring will be done (eg padding tensor will probably be reused on multiple spots, model config will be used instead of some fixed constants like grid, subblock dimensions etc.).

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

File: models/demos/falcon7b/tt/model_utils.py [hunk 1]
```diff
@@ -57,3 +59,88 @@ def get_weights_cached(
...
+    # pad activations to inner dim 144
+    padding = torch.zeros([1, 1, seq_len, 64])
+    padding_t = (
+        tt_lib.tensor.Tensor(padding, in0_dtype)
+        .to(tt_lib.tensor.Layout.TILE)
+        .to(hidden_states.device(), in0_mem_config)
+    )
+    hidden_states = tt_lib.tensor.concat([hidden_states, padding_t], -1)
+
+    compute_kernel_config = tt_lib.tensor.WormholeComputeKernelConfig(
```
Comment (miacim) [inline models/demos/falcon7b/tt/model_utils.py:100]: Since this function supports only wormhole, it would be good to prevent usage of it by Grayskull demo/tests

File: models/demos/falcon7b/tests/unit_tests/test_falcon_lm_head_matmul_2d.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    # calculate parameters for the given sequence length
+    num_slices = 4 if seq_len <= 1024 else 8
+
+    # --- prepare activations
+    A = torch.randn(a_shape)
+    padding = torch.zeros([1, 1, M * 32, 64])
+    A_padded = torch.cat([A, padding], -1)
+
+    # padd in0 tensor to have last dimension 144 * 32 instead of 142 * 32
+    a_t = ttl.tensor.Tensor(A, in0_dtype).to(ttl.tensor.Layout.TILE).to(device, in0_mem_config)
```
Comment (pavlejosipovic) [inline models/demos/falcon7b/tests/unit_tests/test_falcon_lm_head_matmul_2d.py:39]: I've seen tt_lib usually using torch2tt_tensor util function to get tensor from torch to device

File: models/demos/falcon7b/tests/unit_tests/test_falcon_lm_head_matmul_2d.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    # --- prepare weights
+    B = torch.randn(b_shape) - 0.95
```
Comment (pavlejosipovic) [inline models/demos/falcon7b/tests/unit_tests/test_falcon_lm_head_matmul_2d.py:42]: why -0.95?

File: models/demos/falcon7b/tt/model_utils.py [hunk 4]
```diff
@@ -57,3 +59,88 @@ def get_weights_cached(
...
+        math_fidelity=tt_lib.tensor.MathFidelity.LoFi,
+        math_approx_mode=True,
+        fp32_dest_acc_en=False,
+        packer_l1_acc=True,
+    )
+
+    grid = (8, 8)
```
Comment (pavlejosipovic) [inline models/demos/falcon7b/tt/model_utils.py:107]: this won't fit ATM on nebula x2, is this a problem?

**Review by pavlejosipovic — APPROVED** [review]

**Review by pavlepopovic — APPROVED** [review]

**Review by djordje-tt — APPROVED** [review]
